### PR TITLE
ORCA-4515: GitHub Checks Initial Code Path

### PR DESCRIPTION
**Pull request by pcalley**
Adds scaffolding for using Github Checks instead of Comments.

1. Feature aware PullUpdater and CommitStatusUpdater toggle between Github Checks if the feature is enabled and Comments/Commit Status'.
2. CheckID field is used to update the existing Check run when the results of the command are ready.

### CODE REVIEW

File: server/events/plan_command_runner.go [hunk 1]
```diff
@@ -101,7 +103,8 @@ func (p *PlanCommandRunner) runAutoplan(ctx *command.Context) {
 		return
 	}
 
-	// At this point we are sure Atlantis has work to do, so set commit status to pending
+	// At this point we are sure Atlantis has work to do,
+	// so set commit status to pending and store the checkId to be updated.
```
Comment (nishkrishnan) [inline server/events/plan_command_runner.go:107]: let's revert this comment back to what it was

File: server/events/pull_updater.go [hunk 2]
```diff
@@ -41,3 +47,50 @@ func (c *PullUpdater) updatePull(ctx *command.Context, cmd PullCommand, res comm
...
+	VCSClient            vcs.Client
+	MarkdownRenderer     *MarkdownRenderer
+	GlobalCfg            valid.GlobalCfg
+	FeatureAllocator     feature.Allocator
+}
+
+func (f *FeatureAwarePullUpdater) updatePull(ctx *command.Context, cmd PullCommand, res command.Result) {
+	githubChecks, err := f.FeatureAllocator.ShouldAllocate(feature.GitHubChecks, ctx.HeadRepo.FullName)
+	if err != nil {
+		githubChecks = false
```
Comment (nishkrishnan) [inline server/events/pull_updater.go:63]: why are we just swallowing the error here? How would we know if something was wrong? We should at least log it.

**Review by nishkrishnan — COMMENTED** [review]
Hmm i think you're on track with part of the approach, but in it's current form I don't think you'd be able to contribute this upstream given that we're adding a lot of github specific things to generic parts of the code.  

Here's what I'm proposing instead.

Use the VCSClient UpdateStatus method to handle the branching between checks and statuses.  You can pass in some sort of options into the function to tell it whether to use checks or statuses, this would be determined by two things: 
1. the feature flag for internal launching
2. a server flag so that OSS community users can enable checks for their usecases.

The UpdateStatus method would also need to return some generic id which github client would use as the check id.

Let me know if you need to discuss offline.

File: server/events/pull_updater.go [hunk 3]
```diff
@@ -41,3 +47,50 @@ func (c *PullUpdater) updatePull(ctx *command.Context, cmd PullCommand, res comm
 		ctx.Log.Err("unable to comment: %s", err)
 	}
 }
+
+type FeatureAwarePullUpdater struct {
```
Comment (nishkrishnan) [inline server/events/pull_updater.go:51]: can you separate out code that is lyft specific vs. upstreamable? ie. lyft specific code should go in a subpackage called lyft, or just in the `server/lyft` package.

See how @msarvar does it here:
https://github.com/lyft/atlantis/pull/197/files

File: server/events/markdown_renderer.go [hunk 4]
```diff
@@ -124,6 +124,11 @@ func (m *MarkdownRenderer) Render(res command.Result, cmdName command.Name, log
 	return m.renderProjectResults(res.ProjectResults, common, vcsHost, templateOverrides)
 }
 
+func (m *MarkdownRenderer) RenderSingleProjectResult(res command.ProjectResult, cmdName command.Name, log string, verbose bool, vcsHost models.VCSHostType, templateOverrides map[string]string) string {
+	// TODO: Implement render single project result in order to support Github checks for individual projects.
+	return ""
```
[the remark this answers is not in the document]
Comment (pcalley) [reply server/events/markdown_renderer.go]: done

File: server/events/pull_updater.go (hunk 3 above)
Comment (pcalley) [reply server/events/pull_updater.go]: done

File: server/events/commit_status_updater.go [hunk 5]
```diff
@@ -38,13 +39,40 @@ type CommitStatusUpdater interface {
...
+}
+
+func (f *FeatureAwareCommitStatusUpdater) UpdateProject(ctx command.ProjectContext, cmdName command.Name, status models.CommitStatus, url string) error {
+	githubChecks, err := f.FeatureAllocator.ShouldAllocate(feature.GitHubChecks, ctx.HeadRepo.FullName)
+	if err != nil {
+		githubChecks = false
+	}
+	if githubChecks {
+		if status == models.PendingCommitStatus {
+			// This is a new commit or comment command so we create a new Check Run.
```
[the remark this answers is not in the document]
Comment (pcalley) [reply server/events/commit_status_updater.go]: yep this is covered as part of creating the Status ID during context creation.

**Pull request closed without merging**